### ISSUE TITLE
Error on Importing bacpac in DEV VM

### ISSUE THREAD

**Issue by blonder-Andi** [d365collaborative/d365fo.tools#511]
Hi,
i am getting an error at importing by using the following command:

> _Import-D365Bacpac -ImportModeTier1 -BacpacFile ‘C:\TEMP\filename.bacpac’ -NewDatabaseName ‘AXDB_YYYYMMDD’ -Verbose_ 

This error message came up:

> __VERBOSE: [09:04:08][Get-SQLCommand] Building the SQL connection string._
> [09:04:08][Set-SqlBacpacValues] Something went wrong while working against the database. | Exception calling "ExecuteNonQuery" with "0"
> argument(s): "'axmrruntimeuser' is not a valid login or you do not have permission._

The reason behind this is, that there is no "axmrruntimeuser" login in my DEV Box SQL Instance.
_

> \internal\sql\set-bacpacvaluessql.sql (line 18)
> CREATE USER axmrruntimeuser FROM LOGIN axmrruntimeuser

_

I don´t have this user on my DEV Boxes.
So can you change the script accordingly to check if the login exists?
Best Regards
Andreas

**Comment by Splaxi** [comment]
Hi,

Thanks reporting this issue.

What version are you running on your Tier1? Is it deployed from LCS or a download of the VHD file from the LCS Asset Library?

I'm trying to make sure that making a change, that helps you, doesn't break it for all the other people that rely on the tools.

You could also simply create the login by hand and that way have the script complete, while you wait for me to fix.

**Comment by Splaxi** [comment]
As a test - could you try and replace the content of the set-bacpacvaluessql.sql file with this:

```
--Author: Rasmus Andersen (@ITRasmus)
--Author: Tommy Skaue (@skaue)

DROP USER IF EXISTS [axretailruntimeuser]
DROP USER IF EXISTS [axretaildatasyncuser]
DROP USER IF EXISTS [axmrruntimeuser]
DROP USER IF EXISTS [axdeployuser]
DROP USER IF EXISTS [axdbadmin]
DROP USER IF EXISTS [axdeployextuser]
DROP USER IF EXISTS [NT AUTHORITY\NETWORK SERVICE]

IF EXISTS (SELECT * FROM sys.syslogins WHERE NAME = 'axdeployuser')
BEGIN
	CREATE USER axdeployuser FROM LOGIN axdeployuser
	EXEC sp_addrolemember 'db_owner', 'axdeployuser'
END

IF EXISTS (SELECT * FROM sys.syslogins WHERE NAME = 'axdbadmin')
BEGIN
	CREATE USER axdbadmin FROM LOGIN axdbadmin
	EXEC sp_addrolemember 'db_owner', 'axdbadmin'
END

IF EXISTS (SELECT * FROM sys.syslogins WHERE NAME = 'axmrruntimeuser')
BEGIN
	CREATE USER axmrruntimeuser FROM LOGIN axmrruntimeuser
	EXEC sp_addrolemember 'db_datareader', 'axmrruntimeuser'
	EXEC sp_addrolemember 'db_datawriter', 'axmrruntimeuser'
END

IF EXISTS (SELECT * FROM sys.syslogins WHERE NAME = 'axretaildatasyncuser')
BEGIN
	CREATE USER axretaildatasyncuser FROM LOGIN axretaildatasyncuser
	IF (DATABASE_PRINCIPAL_ID('DataSyncUsersRole') IS NOT NULL)
	BEGIN
		EXEC sp_addrolemember 'DataSyncUsersRole', 'axretaildatasyncuser'
	END
END

IF EXISTS (SELECT * FROM sys.syslogins WHERE NAME = 'axretailruntimeuser')
BEGIN
	CREATE USER axretailruntimeuser FROM LOGIN axretailruntimeuser
	IF (DATABASE_PRINCIPAL_ID('UsersRole') IS NOT NULL)
	BEGIN
		EXEC sp_addrolemember 'UsersRole', 'axretailruntimeuser'

	END
	
	IF (DATABASE_PRINCIPAL_ID('ReportUsersRole') IS NOT NULL)
	BEGIN
		EXEC sp_addrolemember 'ReportUsersRole', 'axretailruntimeuser'
	END
END

IF EXISTS (SELECT * FROM sys.syslogins WHERE NAME = 'axdeployextuser')
BEGIN
	CREATE USER axdeployextuser FROM LOGIN axdeployextuser
	IF (DATABASE_PRINCIPAL_ID('DeployExtensibilityRole') IS NOT NULL)
	BEGIN
		EXEC sp_addrolemember 'DeployExtensibilityRole', 'axdeployextuser'
	END
END

CREATE USER [NT AUTHORITY\NETWORK SERVICE] FROM LOGIN [NT AUTHORITY\NETWORK SERVICE]
EXEC sp_addrolemember 'db_owner', 'NT AUTHORITY\NETWORK SERVICE'

UPDATE T1
SET T1.storageproviderid = 0
    , T1.accessinformation = ''
    , T1.modifiedby = 'Admin'
    , T1.modifieddatetime = getdate()
FROM docuvalue T1
WHERE T1.storageproviderid = 1 --Azure storage


IF((SELECT 1
FROM SYS.CHANGE_TRACKING_DATABASES
WHERE DATABASE_ID = DB_ID('@DATABASENAME')) IS NULL)
BEGIN
	ALTER DATABASE [@DATABASENAME] SET CHANGE_TRACKING = ON (CHANGE_RETENTION = 6 DAYS, AUTO_CLEANUP = ON)
END

;--GO
DROP PROCEDURE IF EXISTS SP_ConfigureTablesForChangeTracking
DROP PROCEDURE IF EXISTS SP_ConfigureTablesForChangeTracking_V2
;--GO
-- Begin Refresh Retail FullText Catalogs
DECLARE @RFTXNAME NVARCHAR(MAX);
DECLARE @RFTXSQL NVARCHAR(MAX);
DECLARE retail_ftx CURSOR FOR
SELECT OBJECT_SCHEMA_NAME(object_id) + '.' + OBJECT_NAME(object_id) fullname
FROM SYS.FULLTEXT_INDEXES
WHERE FULLTEXT_CATALOG_ID = (SELECT TOP 1
	FULLTEXT_CATALOG_ID
FROM SYS.FULLTEXT_CATALOGS
WHERE NAME = 'COMMERCEFULLTEXTCATALOG');
OPEN retail_ftx;
FETCH NEXT FROM retail_ftx INTO @RFTXNAME;

BEGIN TRY
	WHILE @@FETCH_STATUS = 0
	BEGIN
		PRINT 'Refreshing Full Text Index ' + @RFTXNAME;
		EXEC SP_FULLTEXT_TABLE @RFTXNAME, 'activate';
		SET @RFTXSQL = 'ALTER FULLTEXT INDEX ON ' + @RFTXNAME + ' START FULL POPULATION';
		EXEC SP_EXECUTESQL @RFTXSQL;
		FETCH NEXT FROM retail_ftx INTO @RFTXNAME;
	END
END TRY
BEGIN CATCH
	PRINT error_message()
END CATCH

CLOSE retail_ftx;
DEALLOCATE retail_ftx;
-- End Refresh Retail FullText Catalogs
```

This should test for each login, that we want to create a user for, if it exists. It doesn't fail for a missing login, but it will then skip creating the associated user.

**Comment by blonder-Andi** [comment]
> 
> 
> Hi,
> 
> Thanks reporting this issue.
> 
> What version are you running on your Tier1? Is it deployed from LCS or a download of the VHD file from the LCS Asset Library?
> 
> I'm trying to make sure that making a change, that helps you, doesn't break it for all the other people that rely on the tools.
> 
> You could also simply create the login by hand and that way have the script complete, while you wait for me to fix.

It is a downloaded VHD file from Asset Library. (based on 10.0.13 and upgraded 10.0.15)

**Comment by blonder-Andi** [comment]
> 
> 
> As a test - could you try and replace the content of the set-bacpacvaluessql.sql file with this:
> 
> ```
> --Author: Rasmus Andersen (@ITRasmus)
> --Author: Tommy Skaue (@skaue)
> 
> DROP USER IF EXISTS [axretailruntimeuser]
> DROP USER IF EXISTS [axretaildatasyncuser]
> DROP USER IF EXISTS [axmrruntimeuser]
> DROP USER IF EXISTS [axdeployuser]
> DROP USER IF EXISTS [axdbadmin]
> DROP USER IF EXISTS [axdeployextuser]
> DROP USER IF EXISTS [NT AUTHORITY\NETWORK SERVICE]
> 
> IF EXISTS (SELECT * FROM sys.syslogins WHERE NAME = 'axdeployuser')
> BEGIN
> 	CREATE USER axdeployuser FROM LOGIN axdeployuser
> 	EXEC sp_addrolemember 'db_owner', 'axdeployuser'
> END
> 
> IF EXISTS (SELECT * FROM sys.syslogins WHERE NAME = 'axdbadmin')
> BEGIN
> 	CREATE USER axdbadmin FROM LOGIN axdbadmin
> 	EXEC sp_addrolemember 'db_owner', 'axdbadmin'
> END
> 
> IF EXISTS (SELECT * FROM sys.syslogins WHERE NAME = 'axmrruntimeuser')
> BEGIN
> 	CREATE USER axmrruntimeuser FROM LOGIN axmrruntimeuser
> 	EXEC sp_addrolemember 'db_datareader', 'axmrruntimeuser'
> 	EXEC sp_addrolemember 'db_datawriter', 'axmrruntimeuser'
> END
> 
> IF EXISTS (SELECT * FROM sys.syslogins WHERE NAME = 'axretaildatasyncuser')
> BEGIN
> 	CREATE USER axretaildatasyncuser FROM LOGIN axretaildatasyncuser
> 	IF (DATABASE_PRINCIPAL_ID('DataSyncUsersRole') IS NOT NULL)
> 	BEGIN
> 		EXEC sp_addrolemember 'DataSyncUsersRole', 'axretaildatasyncuser'
> 	END
> END
> 
> IF EXISTS (SELECT * FROM sys.syslogins WHERE NAME = 'axretailruntimeuser')
> BEGIN
> 	CREATE USER axretailruntimeuser FROM LOGIN axretailruntimeuser
> 	IF (DATABASE_PRINCIPAL_ID('UsersRole') IS NOT NULL)
> 	BEGIN
> 		EXEC sp_addrolemember 'UsersRole', 'axretailruntimeuser'
> 
> 	END
> 	
> 	IF (DATABASE_PRINCIPAL_ID('ReportUsersRole') IS NOT NULL)
> 	BEGIN
> 		EXEC sp_addrolemember 'ReportUsersRole', 'axretailruntimeuser'
> 	END
> END
> 
> IF EXISTS (SELECT * FROM sys.syslogins WHERE NAME = 'axdeployextuser')
> BEGIN
> 	CREATE USER axdeployextuser FROM LOGIN axdeployextuser
> 	IF (DATABASE_PRINCIPAL_ID('DeployExtensibilityRole') IS NOT NULL)
> 	BEGIN
> 		EXEC sp_addrolemember 'DeployExtensibilityRole', 'axdeployextuser'
> 	END
> END
> 
> CREATE USER [NT AUTHORITY\NETWORK SERVICE] FROM LOGIN [NT AUTHORITY\NETWORK SERVICE]
> EXEC sp_addrolemember 'db_owner', 'NT AUTHORITY\NETWORK SERVICE'
> 
> UPDATE T1
> SET T1.storageproviderid = 0
>     , T1.accessinformation = ''
>     , T1.modifiedby = 'Admin'
>     , T1.modifieddatetime = getdate()
> FROM docuvalue T1
> WHERE T1.storageproviderid = 1 --Azure storage
> 
> 
> IF((SELECT 1
> FROM SYS.CHANGE_TRACKING_DATABASES
> WHERE DATABASE_ID = DB_ID('@DATABASENAME')) IS NULL)
> BEGIN
> 	ALTER DATABASE [@DATABASENAME] SET CHANGE_TRACKING = ON (CHANGE_RETENTION = 6 DAYS, AUTO_CLEANUP = ON)
> END
> 
> ;--GO
> DROP PROCEDURE IF EXISTS SP_ConfigureTablesForChangeTracking
> DROP PROCEDURE IF EXISTS SP_ConfigureTablesForChangeTracking_V2
> ;--GO
> -- Begin Refresh Retail FullText Catalogs
> DECLARE @RFTXNAME NVARCHAR(MAX);
> DECLARE @RFTXSQL NVARCHAR(MAX);
> DECLARE retail_ftx CURSOR FOR
> SELECT OBJECT_SCHEMA_NAME(object_id) + '.' + OBJECT_NAME(object_id) fullname
> FROM SYS.FULLTEXT_INDEXES
> WHERE FULLTEXT_CATALOG_ID = (SELECT TOP 1
> 	FULLTEXT_CATALOG_ID
> FROM SYS.FULLTEXT_CATALOGS
> WHERE NAME = 'COMMERCEFULLTEXTCATALOG');
> OPEN retail_ftx;
> FETCH NEXT FROM retail_ftx INTO @RFTXNAME;
> 
> BEGIN TRY
> 	WHILE @@FETCH_STATUS = 0
> 	BEGIN
> 		PRINT 'Refreshing Full Text Index ' + @RFTXNAME;
> 		EXEC SP_FULLTEXT_TABLE @RFTXNAME, 'activate';
> 		SET @RFTXSQL = 'ALTER FULLTEXT INDEX ON ' + @RFTXNAME + ' START FULL POPULATION';
> 		EXEC SP_EXECUTESQL @RFTXSQL;
> 		FETCH NEXT FROM retail_ftx INTO @RFTXNAME;
> 	END
> END TRY
> BEGIN CATCH
> 	PRINT error_message()
> END CATCH
> 
> CLOSE retail_ftx;
> DEALLOCATE retail_ftx;
> -- End Refresh Retail FullText Catalogs
> ```
> 
> This should test for each login, that we want to create a user for, if it exists. It doesn't fail for a missing login, but it will then skip creating the associated user.

I retested with the given changes and the import worked flawlessly.

D365 is working in DEV Box with restored data.

**Comment by Splaxi** [comment]
Should be solved in 0.6.52, which is available on powershellgallery.com.

Please, if you can find time to update the tools, and run it without changing the file and confirm it works, it would be of great help for us!
